File: docscan/web/js/enhance.js
/* Image clean-up - a port of docscan/enhance.py.
 *
 * One deliberate difference: the illumination background is estimated on a
 * downscaled copy and scaled back up.  The background is low frequency by
 * definition, so the result is the same, but it avoids running a large median
 * filter over a full resolution photo, which is painfully slow in WebAssembly.
 */

import { withMats } from "./cv.js";
import { applyLut, mergeChannels, odd, percentile8u, splitChannels, toGray } from "./mat.js";

export const MODES = ["color", "magic", "gray", "bw", "none"];

export const MODE_LABELS = {
  color: "컬러",
  magic: "선명",
  gray: "회색",
  bw: "흑백",
  none: "원본",
};

export const MODE_DESCRIPTIONS = {
  color: "조명을 고르게 펴고 흰 배경을 살린 자연스러운 컬러",
  magic: "대비와 채도를 올린 스캐너 앱 느낌",
  gray: "회색조, 인쇄와 용량에 유리",
  bw: "순수 흑백, 글자 문서에 가장 작은 파일",
  none: "보정 없이 기울기만 보정",
};

const BACKGROUND_SIZE = 512;

/** Estimate the illumination of one channel by growing its bright areas. */
function backgroundOf(channel, kernelSize, blurSize) {
  const cv = window.cv;
  return withMats((keep) => {
    const kernel = keep(cv.getStructuringElement(cv.MORPH_ELLIPSE,
                                                 new cv.Size(kernelSize, kernelSize)));
    const dilated = keep(new cv.Mat());
    cv.dilate(channel, dilated, kernel);
    const background = new cv.Mat();
    cv.medianBlur(dilated, background, blurSize);
    return background;
  });
}

/** Flatten uneven lighting by dividing the image by its own background. */
export function removeShadows(image, strength = 1) {
  const cv = window.cv;
  if (strength <= 0) return image.clone();
  const longest = Math.max(image.rows, image.cols);
  const scale = Math.min(1, BACKGROUND_SIZE / longest);
  const smallSize = new cv.Size(Math.max(16, Math.round(image.cols * scale)),
                                Math.max(16, Math.round(image.rows * scale)));
  const kernelSize = odd(Math.max(image.cols, image.rows) * scale * 0.012, 3);
  const blurSize = odd(Math.max(image.cols, image.rows) * scale * 0.02, 3);

  const channels = splitChannels(image);
  const flattened = [];
  try {
    for (const channel of channels) {
      flattened.push(withMats((keep) => {
        const small = keep(new cv.Mat());
        cv.resize(channel, small, smallSize, 0, 0, cv.INTER_AREA);
        const smallBackground = keep(backgroundOf(small, kernelSize, blurSize));
        const background = keep(new cv.Mat());
        cv.resize(smallBackground, background, new cv.Size(image.cols, image.rows),
                  0, 0, cv.INTER_LINEAR);
        const normalised = new cv.Mat();
        cv.divide(channel, background, normalised, 255);
        if (strength < 1) {
          const blended = new cv.Mat();
          cv.addWeighted(normalised, strength, channel, 1 - strength, 0, blended);
          normalised.delete();
          return blended;
        }
        return normalised;
      }));
    }
    return flattened.length === 1 ? flattened[0].clone() : mergeChannels(flattened);
  } finally {
    channels.forEach((channel) => channel.delete());
    if (flattened.length !== 1) flattened.forEach((channel) => channel.delete());
    else flattened[0].delete();
  }
}

/** Push the darkest and brightest percentiles to pure black and white. */
export function stretchLevels(image, lowPercentile = 1, highPercentile = 99) {
  const cv = window.cv;
  return withMats((keep) => {
    const reference = keep(toGray(image));
    const low = percentile8u(reference, lowPercentile);
    const high = percentile8u(reference, highPercentile);
    if (high - low < 1) return image.clone();
    const scale = 255 / (high - low);
    const result = new cv.Mat();
    image.convertTo(result, -1, scale, -low * scale);
    return result;
  });
}

/** Map near white pixels to pure white with a smooth lightness curve. */
export function whiten(image, whitePoint = 225) {
  const cv = window.cv;
  const point = Math.min(255, Math.max(32, Math.round(whitePoint)));
  const table = new Uint8Array(256);
  for (let value = 0; value < 256; value += 1) {
    table[value] = Math.min(255, Math.round((value * 255) / point));
  }
  if (image.channels() === 1) return applyLut(image, Array.from(table));
  return withMats((keep) => {
    const lab = keep(new cv.Mat());
    cv.cvtColor(image, lab, cv.COLOR_RGB2Lab);
    const channels = splitChannels(lab).map((channel) => keep(channel));
    const lightness = keep(applyLut(channels[0], Array.from(table)));
    const merged = keep(mergeChannels([lightness, channels[1], channels[2]]));
    const result = new cv.Mat();
    cv.cvtColor(merged, result, cv.COLOR_Lab2RGB);
    return result;
  });
}

export function unsharpMask(image, amount = 0.6, radius = 1.4) {
  const cv = window.cv;
  if (amount <= 0) return image.clone();
  return withMats((keep) => {
    const blurred = keep(new cv.Mat());
    cv.GaussianBlur(image, blurred, new cv.Size(0, 0), radius);
    const result = new cv.Mat();
    cv.addWeighted(image, 1 + amount, blurred, -amount, 0, result);
    return result;
  });
}

/** Scale each channel so that its bright pixels become neutral white. */
export function autoWhiteBalance(image, percent = 97) {
  const cv = window.cv;
  if (image.channels() < 3) return image.clone();
  const channels = splitChannels(image);
  const balanced = [];
  try {
    for (const channel of channels) {
      const reference = percentile8u(channel, percent);
      const gain = reference > 1 ? Math.min(255 / reference, 3) : 1;
      const scaled = new cv.Mat();
      channel.convertTo(scaled, -1, gain, 0);
      balanced.push(scaled);
    }
    return mergeChannels(balanced);
  } finally {
    channels.forEach((channel) => channel.delete());
    balanced.forEach((channel) => channel.delete());
  }
}

export function boostSaturation(image, factor = 1.15) {
  const cv = window.cv;
  if (image.channels() < 3 || factor === 1) return image.clone();
  return withMats((keep) => {
    const hsv = keep(new cv.Mat());
    cv.cvtColor(image, hsv, cv.COLOR_RGB2HSV);
    const channels = splitChannels(hsv).map((channel) => keep(channel));
    const saturation = keep(new cv.Mat());
    channels[1].convertTo(saturation, -1, factor, 0);
    const merged = keep(mergeChannels([channels[0], saturation, channels[2]]));
    const result = new cv.Mat();
    cv.cvtColor(merged, result, cv.COLOR_HSV2RGB);
    return result;
  });
}

/** CLAHE on the lightness channel only, so colours stay intact. */
export function localContrast(image, clipLimit = 2, tile = 8) {
  const cv = window.cv;
  const clahe = new cv.CLAHE(clipLimit, new cv.Size(tile, tile));
  try {
    if (image.channels() === 1) {
      const result = new cv.Mat();
      clahe.apply(image, result);
      return result;
    }
    return withMats((keep) => {
      const lab = keep(new cv.Mat());
      cv.cvtColor(image, lab, cv.COLOR_RGB2Lab);
      const channels = splitChannels(lab).map((channel) => keep(channel));
      const lightness = keep(new cv.Mat());
      clahe.apply(channels[0], lightness);
      const merged = keep(mergeChannels([lightness, channels[1], channels[2]]));
      const result = new cv.Mat();
      cv.cvtColor(merged, result, cv.COLOR_Lab2RGB);
      return result;
    });
  } finally {
    clahe.delete();
  }
}

export function toGrayscale(image, shadow = 1) {
  return withMats((keep) => {
    const gray = keep(toGray(image));
    const flattened = keep(removeShadows(gray, shadow));
    const stretched = keep(stretchLevels(flattened));
    return whiten(stretched, 235);
  });
}

/** Crisp bi-level page; the parameters match the measured optimum in Python. */
export function toBlackWhite(image, shadow = 1, offset = 10, blockRatio = 0.025) {
  const cv = window.cv;
  return withMats((keep) => {
    const gray = keep(toGray(image));
    const flattened = keep(removeShadows(gray, shadow));
    const blockSize = odd(Math.max(image.rows, image.cols) * blockRatio, 15);
    const binary = keep(new cv.Mat());
    cv.adaptiveThreshold(flattened, binary, 255, cv.ADAPTIVE_THRESH_GAUSSIAN_C,
                         cv.THRESH_BINARY, blockSize, offset);
    // remove the speckles the threshold leaves in empty areas
    const denoised = new cv.Mat();
    cv.medianBlur(binary, denoised, 3);
    return denoised;
  });
}

export function toMagicColour(image, shadow = 1) {
  return withMats((keep) => {
    const flattened = keep(removeShadows(image, shadow));
    const balanced = keep(autoWhiteBalance(flattened));
    const stretched = keep(stretchLevels(balanced, 2, 99.5));
    const contrasted = keep(localContrast(stretched, 1.3));
    const whitened = keep(whiten(contrasted, 225));
    const saturated = keep(boostSaturation(whitened, 1.2));
    return unsharpMask(saturated, 0.5);
  });
}

export function toColour(image, shadow = 1) {
  return withMats((keep) => {
    const flattened = keep(removeShadows(image, shadow));
    const balanced = keep(autoWhiteBalance(flattened, 98));
    const whitened = keep(whiten(balanced, 240));
    return unsharpMask(whitened, 0.3);
  });
}

/** Apply a colour mode to an already rectified page; returns a new Mat. */
export function enhance(image, mode = "color", shadow = 1) {
  if (!MODES.includes(mode)) throw new Error("unknown mode: " + mode);
  if (mode === "none") return image.clone();
  if (mode === "gray") return toGrayscale(image, shadow);
  if (mode === "bw") return toBlackWhite(image, shadow);
  if (mode === "magic") return toMagicColour(image, shadow);
  return toColour(image, shadow);
}

File: docscan/enhance.py
"""Image clean-up: shadow removal, colour modes and sharpening.

Copyright 2026 PyCAM contributors

This file is part of the docscan document scanner.  It is free software:
you can redistribute it and/or modify it under the terms of the GNU General
Public License as published by the Free Software Foundation, either version 3
of the License, or (at your option) any later version.
"""

import numpy as np

import cv2


#: available output looks, in the order used by the "cycle mode" key of the camera UI
MODES = ("color", "magic", "gray", "bw", "none")

MODE_DESCRIPTIONS = {
    "none": "no processing, only the perspective correction",
    "color": "colour photo with even lighting and a white background",
    "magic": "boosted colour document (the classic scanner app look)",
    "gray": "neutral grayscale",
    "bw": "hard black and white, smallest files, best for plain text",
}


def _odd(value, minimum=3):
    value = int(round(value))
    if value % 2 == 0:
        value += 1
    return max(minimum, value)


def _background(channel, kernel_size, blur_size):
    """Estimate the illumination of a channel by growing its bright areas."""
    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (kernel_size, kernel_size))
    dilated = cv2.dilate(channel, kernel)
    return cv2.medianBlur(dilated, blur_size)


def remove_shadows(image, strength=1.0):
    """Flatten uneven lighting by dividing the image by its own background.

    This is what turns a photo taken under a desk lamp into something that
    looks like it came out of a flatbed scanner.
    """
    if strength <= 0:
        return image
    longest = max(image.shape[:2])
    kernel_size = _odd(longest * 0.012, 3)
    blur_size = _odd(longest * 0.02, 3)
    if image.ndim == 2:
        channels = [image]
    else:
        channels = cv2.split(image)
    flattened = []
    for channel in channels:
        background = _background(channel, kernel_size, blur_size)
        normalised = cv2.divide(channel, background, scale=255)
        if strength < 1.0:
            normalised = cv2.addWeighted(normalised, strength, channel, 1.0 - strength, 0)
        flattened.append(normalised)
    if len(flattened) == 1:
        return flattened[0]
    return cv2.merge(flattened)


def stretch_levels(image, low_percentile=1.0, high_percentile=99.0):
    """Push the darkest and brightest percentiles to pure black and white."""
    result = image.astype(np.float32)
    reference = image if image.ndim == 2 else cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    low = float(np.percentile(reference, low_percentile))
    high = float(np.percentile(reference, high_percentile))
    if high - low < 1e-3:
        return image
    result = (result - low) * (255.0 / (high - low))
    return np.clip(result, 0, 255).astype(np.uint8)


def unsharp_mask(image, amount=0.6, radius=1.4):
    """Sharpen edges without amplifying noise too much."""
    if amount <= 0:
        return image
    blurred = cv2.GaussianBlur(image, (0, 0), radius)
    return cv2.addWeighted(image, 1.0 + amount, blurred, -amount, 0)


def auto_white_balance(image, percentile=97.0):
    """Scale each channel so that its bright pixels become neutral white."""
    if image.ndim != 3:
        return image
    channels = []
    for channel in cv2.split(image):
        reference = float(np.percentile(channel, percentile))
        gain = 255.0 / reference if reference > 1.0 else 1.0
        gain = min(gain, 3.0)
        channels.append(np.clip(channel.astype(np.float32) * gain, 0, 255).astype(np.uint8))
    return cv2.merge(channels)


def boost_saturation(image, factor=1.15):
    if image.ndim != 3 or factor == 1.0:
        return image
    hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV).astype(np.float32)
    hsv[:, :, 1] = np.clip(hsv[:, :, 1] * factor, 0, 255)
    return cv2.cvtColor(hsv.astype(np.uint8), cv2.COLOR_HSV2BGR)


def whiten(image, white_point=225):
    """Map near-white pixels to pure white with a smooth lightness curve.

    Paper is never perfectly uniform, and the local contrast step happily
    amplifies that leftover texture.  Pulling everything above ``white_point``
    up to white removes it again without touching the ink.
    """
    white_point = int(np.clip(white_point, 32, 255))
    table = np.clip(np.arange(256, dtype=np.float32) * (255.0 / white_point),
                    0, 255).astype(np.uint8)
    if image.ndim == 2:
        return cv2.LUT(image, table)
    lab = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
    lab[:, :, 0] = cv2.LUT(lab[:, :, 0], table)
    return cv2.cvtColor(lab, cv2.COLOR_LAB2BGR)


def local_contrast(image, clip_limit=2.0, tile=8):
    """Apply CLAHE on the lightness channel only, so colours stay intact."""
    clahe = cv2.createCLAHE(clipLimit=clip_limit, tileGridSize=(tile, tile))
    if image.ndim == 2:
        return clahe.apply(image)
    lab = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
    lab[:, :, 0] = clahe.apply(lab[:, :, 0])
    return cv2.cvtColor(lab, cv2.COLOR_LAB2BGR)


def to_grayscale(image, shadow=1.0):
    gray = image if image.ndim == 2 else cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    gray = remove_shadows(gray, shadow)
    return whiten(stretch_levels(gray), 235)


def to_black_white(image, shadow=1.0, offset=10, block_ratio=0.025, denoise=True):
    """Produce a crisp bi-level page using a locally adaptive threshold.

    The parameters were picked by measuring how much of the text survives (and
    how much speckle is added) on rendered pages at several resolutions and
    noise levels: blurring first costs thin strokes, while the median filter
    afterwards is what keeps a noisy photo from turning into confetti.
    """
    gray = image if image.ndim == 2 else cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    gray = remove_shadows(gray, shadow)
    block_size = _odd(max(image.shape[:2]) * block_ratio, 15)
    binary = cv2.adaptiveThreshold(gray, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
                                   cv2.THRESH_BINARY, block_size, offset)
    if denoise:
        # remove isolated speckles created by the threshold in empty areas
        binary = cv2.medianBlur(binary, 3)
    return binary


def to_magic_colour(image, shadow=1.0):
    """The typical scanner app look: white paper, saturated ink, crisp edges."""
    result = remove_shadows(image, shadow)
    result = auto_white_balance(result)
    result = stretch_levels(result, 2.0, 99.5)
    result = local_contrast(result, clip_limit=1.3)
    result = whiten(result, 225)
    result = boost_saturation(result, 1.2)
    return unsharp_mask(result, amount=0.5)


def to_colour(image, shadow=1.0):
    """A faithful colour scan: even lighting, neutral white, no heavy grading."""
    result = remove_shadows(image, shadow)
    result = auto_white_balance(result, percentile=98.0)
    result = whiten(result, 240)
    return unsharp_mask(result, amount=0.3)


def enhance(image, mode="color", shadow=1.0, sharpen=None):
    """Apply the requested colour mode to an already rectified page."""
    if mode not in MODES:
        raise ValueError("unknown mode {!r}, expected one of {}".format(mode, ", ".join(MODES)))
    if mode == "none":
        result = image
    elif mode == "gray":
        result = to_grayscale(image, shadow)
    elif mode == "bw":
        result = to_black_white(image, shadow)
    elif mode == "magic":
        result = to_magic_colour(image, shadow)
    else:
        result = to_colour(image, shadow)
    if sharpen and mode != "bw":
        result = unsharp_mask(result, amount=sharpen)
    return result
